exercises/matrix_determinant/python/main.py

```diff
@@ -0,0 +1,64 @@
+from copy import copy, deepcopy
+
+def get_dim():
+    # Get number of rows and columns from user
+    while (True):
+        try:
+            dim = int(input("Enter number of rows/columns: "))
+            if (dim <= 0):
+                raise ValueError
+            break
+        except ValueError:
+            print("Please enter positive integer")
+    return dim
+
+# Gets a real-valued square matrix from the user
+def get_square_matrix():
+    dim = get_dim()
+
+    # Get matrix values
+    matrix = []
+    print("Enter entries rowwise with enters in between: ")
+
+    for i in range(dim):
+        print("Row %d:" %(i+1))
+        row = []
+        while (len(row) < dim):
+            try:
+                row.append(float(input()))
+            except ValueError:
+                print("Please enter real numbers")
+        matrix.append(row)
+
+    return matrix
+
+# Computer the determinant of a real-valued square matrix
+def det(matrix):
+    dim = len(matrix)
+    if (dim == 1):
+        return matrix[0][0]
+    else:
+        ret = 0
+        for i in range(dim):
+            ret += ((-1) ** i) * matrix[0][i] * det(first_minor(matrix,0,i))
+        return ret
+
+# Returns a copy of the inputted matrix with the specified row and column removed
+def first_minor(matrix,r,c):
+    matrix_copy = deepcopy(matrix)
```

Comment:
This works, and is sufficient for our purposes, but if you wanted an extra challenge this would be the best place to optimize your function since this is quite an expensive operation to be doing this many times and it is possible to do without (if you wanted to do this, lmk if you have any questions about how to do this)

Reply:
To avoid copying the matrix, I can pass and edit the original matrix using pointers?

Reply:
You could make a data structure to keep track of which columns and rows to ignore and just pass the original matrix around with this meta data structure.

So for example `[[1,2,3], [4,5,6], [7,8,9]]` with `[(1, 2)]` should be interpreted as `[[1, 2], [7, 8]]` by your functions.

Reply:
At each step, we take each element of the first row and multiply it by the determinant of its first minor. For each of these multiplications, don't we need to make a copy of the original matrix? First example, we cannot "edit" the matrix when multiplying the first element of the row by the determinant of its first minor because we need the original matrix to multiply the second element of the first row by the determinant of its first minor.

Does that make sense?